bin/run_controller

```diff
@@ -0,0 +1,170 @@
+
+docker_volumes=
+port_opts=
+local=
+faucet_id=
+
+cd $ROOT
+
+if [ "$1" == "local" ]; then
+    docker_volumes+=" -v $ROOT/faucet/faucet:/usr/lib/python3.7/site-packages/faucet"
+    local=local
+    shift
+fi
+
+if [ -n "$1" ]; then
+    CONTAINER=$CONTAINER-$1
+    INSTDIR=$INSTDIR/forch-controller-$1
+    faucet_id=$1
+    shift
+fi
+
+if [ -n "$1" ]; then
+    ext_ofpt=$1
+    shift
+fi
+
+if [ -n "$1" ]; then
+    prom_pt=$1
+    shift
+fi
+
+if [ -n "$1" ]; then
+    gauge_ext_ofpt=$1
+    shift
+fi
+
+if [ -n "$1" ]; then
+    gauge_prom_pt=$1
+    shift
+fi
+
+echo Creating $IMAGE instance $CONTAINER on port $ext_ofpt in $INSTDIR
+
+# Check that image exists so docker doesn't pull down random one from dockerhub
+docker inspect $IMAGE > /dev/null
+
+mkdir -p $INSTDIR
+
+if docker container inspect $CONTAINER --format '{{ .Name }}' > /dev/null 2>&1; then
+    echo -n "Clensing old container "
+    docker rm -f $CONTAINER
+fi
+
+if [ -n "$FAUCET_SOCK" ]; then
+    rm -f $INSTDIR/$FAUCET_SOCK
+    event_sock=/var/log/faucet/$FAUCET_SOCK
+fi
+
+faucet_config=/etc/faucet/$FAUCET_CONFIG
+
+if [ -n "$local" ]; then
+    using_faucet=$(cd faucet; git rev-list -n 1 `git describe --dirty`) || using_faucet=dev
+    error_label="from local faucet/"
+    echo Using local faucet version $using_faucet
+else
+    using_faucet=`docker inspect $IMAGE | jq -r .[0].Config.Labels.FAUCET_HASH`
+    error_label="from image $IMAGE"
+
+    expected_faucet=$(cat etc/FAUCET_VERSION)
+
+    using_hash=`cd faucet; git rev-list -n 1 $using_faucet`
+    expected_hash=`cd faucet; git rev-list -n 1 $expected_faucet`
+    if [ "$using_hash" != "$expected_hash" ]; then
+	echo Image faucet version does not match configuration.
+	echo $using_hash $error_label $using_faucet
+	echo $expected_hash from etc/FAUCET_VERSION $expected_faucet
+	false
+    fi
+fi
+
+docker_envs="\
+        --env FAUCET_CONFIG_STAT_RELOAD=1 \
+        --env GAUGE_CONFIG_STAT_RELOAD=1 \
+        --env FAUCET_EVENT_SOCK_HEARTBEAT=10 \
+        --env USERID=$(id -u) \
+        --env FAUCET_EVENT_SOCK=$event_sock \
+        --env FAUCET_CONFIG=$faucet_config \
+        --env FAUCET_LISTEN_PORT=6653 \
+        --env FAUCET_RYU_CONF=/etc/faucet/ryu.conf"
+
+docker_volumes+="\
+        -v $INSTDIR/faucet:/etc/faucet \
+        -v $INSTDIR:/var/log/faucet"
+
+
+if [ -n "$prom_pt" ]; then
+    port_opts="-p $prom_pt:$PROM_TARGET"
+    echo $prom_pt > $INSTDIR/PROM_PORT.$PROM_TARGET
+fi
+
+if [ -n "$gauge_prom_pt" ]; then
+    gauge_port_opts="-p $gauge_prom_pt:$GAUGE_PROM_TARGET"
+    echo $gauge_prom_pt > $INSTDIR/PROM_PORT.$GAUGE_PROM_TARGET
+fi
+
+forch_port="-p 500$faucet_id:9019"
+docker_envs+="\
+    --env CONTROLLER_NAME=$CONTAINER \
+    --env PROMETHEUS_PORT=9302 \
+    --env FAUCET_EVENT_SOCK=/var/log/faucet/faucet_event.sock \
+    --env FAUCET_CONFIG_DIR=/etc/faucet \
+    --env FORCH_CONFIG_DIR=/etc/faucet \
+    --env WSAPI_LISTEN_HOST=localhost \
+    --env WSAPI_LISTEN_PORT=4000 \
```

Comment:
likewise, this is starting to get to the point where it will be a lot cleaner to have all this in a file that's then dynamically loaded into the env at runtime.  Again, not something required to fix now but to ponder going forward.

Reply:
makes sense.